components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt

```diff
@@ -1106,6 +1106,8 @@ class GeckoEngineSession(
         geckoSession.historyDelegate = createHistoryDelegate()
         geckoSession.mediaSessionDelegate = GeckoMediaSessionDelegate(this)
         geckoSession.scrollDelegate = createScrollDelegate()
+
+        runtime.orientationController.delegate = GeckoOrientationDelegate(this)
```

Comment:
Doesn't this mean that the last _created_ GeckoSession will only have the delegate attached?

1. Create a GeckoSession A (delegate attached).
2. Create a GeckoSession B (new B delegate is attached).
3. Navigate to session A.
4. Session A calls orientation change.
5. Nothing would happen?

Reply:
> Doesn't this mean that the last _created_ GeckoSession will only have the delegate attached?

Unfortunately yes.
I tested with multiple tabs and the scenario above works as expected (because it doesn't really matter which session send the orientation change request) - even allowed to found that blocker - https://bugzilla.mozilla.org/show_bug.cgi?id=1757431 
But the issue is that if I'm linking the delegate to `GeckoSession` and the tab for it gets closed I'd loose the delegate.

Really tried to keep this functionality in AC and implement everything in a reactive way so that's why I tried using the `GeckoSession` and a `Middleware` for `BrowserStore` but since the runtime outlives a session this doesn't seem like a good approach.

Because the `GeckoRuntime` is a singleton initialized in Fenix which needs a delegate based on which we'd update the current activity seems like it would make more sense to have everything set in Fenix - each AC / GV consumer? With still an intricate way of calling an activity from a singleton. And then the apps would have to respond with a `GeckoResult` and directly handle GeckoView callbacks, which I don't think they should do.

So I'm wondering where else in AC can this runtime delegate be set? And how to easily communicate changes GeckoView changes / callback to the app.
`GeckoEngine` might be a candidate but not sure each engine needs to set this delegate and to pass a configuration to it to just set the delegate there doesn't seem like a big improvement.

Thinking about the "implement directly in Fenix" approach I guess an activity can set itself as the runtime delegate in `onCreate` and everything would be much simpler but this seems to me like sidestepping all the current architecture layers and communication flow while also bringing GeckoView's `GeckoResult` in the activity 

@jonalmeida  @csadilek Thoughts?

Reply:
@Mugurell @jonalmeida 

Oh I missed in my first pass that this delegate is on the runtime! I think in that case we should API to the `GeckoEngine` in A-C to register a delegate and receive calbacks. We do this already see `WebExtensionDelegate`. Then consumers (Fenix/Focus) could register the A-C delegate and get called for `onOrientationLock` and `onOrientationUnlock`. We can hook up a delegate by default too that dispatches the actions (same as below). They are already independent of the session anyway.

Reply:
I was constructing a reply similar to Christian's, but instead I'll add that there is a reference with WebPush where the engine integrates with other pieces of code ([`WebPushEngineIntegration`](https://github.com/mozilla-mobile/fenix/blob/main/app/src/main/java/org/mozilla/fenix/push/WebPushEngineIntegration.kt)).